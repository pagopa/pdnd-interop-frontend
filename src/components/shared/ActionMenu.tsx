import React, { useId } from 'react'
import { ClickAwayListener, MenuItem, Menu, IconButton, Skeleton, Box } from '@mui/material'
import type { ActionItem, ExtendedMUIColor } from '@/types/common.types'
import MoreVertIcon from '@mui/icons-material/MoreVert'
import { useTranslation } from 'react-i18next'

type ActionMenuProps = {
  actions: Array<ActionItem>
  iconColor?: ExtendedMUIColor
}

export const ActionMenu: React.FC<ActionMenuProps> = ({ actions, iconColor = 'primary' }) => {
  const [open, setOpen] = React.useState(false)
  const anchorRef = React.useRef<HTMLButtonElement>(null)
  const compositionButtonId = useId() + '-composition-button'
  const compositionMenuId = useId() + '-composition-menu'
  const { t } = useTranslation('shared-components', { keyPrefix: 'actionMenu' })

  const handleToggle = () => {
    setOpen((prevOpen) => !prevOpen)
  }

  const handleClose = (event: Event | React.SyntheticEvent) => {
    if (anchorRef.current && anchorRef.current.contains(event.target as HTMLElement)) {
      return
    }

    setOpen(false)
  }

  function handleListKeyDown(event: React.KeyboardEvent) {
    if (event.key === 'Tab') {
      event.preventDefault()
      setOpen(false)
    } else if (event.key === 'Escape') {
      setOpen(false)
    }
  }

  // return focus to the button when we transitioned from !open -> open
  const prevOpen = React.useRef(open)
  React.useEffect(() => {
    if (prevOpen.current === true && open === false) {
      anchorRef.current!.focus()
    }

    prevOpen.current = open
  }, [open])

  return (
    <>
      <IconButton
        ref={anchorRef}
        id={compositionButtonId}
        sx={{ visibility: actions.length > 0 ? 'visible' : 'hidden' }}
        aria-controls={open ? compositionMenuId : undefined}
        aria-expanded={open ? 'true' : undefined}
        aria-haspopup="true"
        onClick={handleToggle}
        aria-label={t('iconButtonAriaLabel')}
      >
        <MoreVertIcon color={iconColor} fontSize="small" />
      </IconButton>
      <ClickAwayListener onClickAway={handleClose}>
        <Menu
          open={open}
          id={compositionMenuId}
          anchorEl={anchorRef.current}
          aria-labelledby={compositionButtonId}
          onClose={handleClose}
          onKeyDown={handleListKeyDown}
          MenuListProps={{
            'aria-labelledby': 'basic-button',
          }}
          anchorOrigin={{ vertical: 'bottom', horizontal: 'right' }}
          transformOrigin={{ vertical: 'top', horizontal: 'right' }}
        >
          {actions.map(({ label, action }) => (
            <MenuItem
              key={label}
              onClick={(e) => {
                action()
                handleClose(e)
              }}
            >
              {label}
            </MenuItem>
          ))}
        </Menu>
      </ClickAwayListener>
    </>
  )
}

export const ActionMenuSkeleton: React.FC = () => {
  return (
    <Box component="span" sx={{ ml: 2, display: 'inline-block' }}>
      <Skeleton variant="rectangular" sx={{ my: 1, mx: 2 }} width={4} />
    </Box>
  )
}
