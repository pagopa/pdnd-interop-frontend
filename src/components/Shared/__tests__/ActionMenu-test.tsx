import React from 'react'
import renderer from 'react-test-renderer'
import noop from 'lodash/noop'
import { ActionMenu } from '../ActionMenu'
// import { AllTheProviders } from '../../../__mocks__/providers'

describe('Snapshot test', () => {
  it('renders null with no actions', () => {
    const component = renderer.create(<ActionMenu index={0} actions={[]}></ActionMenu>)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders button', () => {
    const actions = [{ onClick: noop, label: 'Azione 1' }]
    const component = renderer.create(<ActionMenu index={0} actions={actions}></ActionMenu>)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  // it('renders list', () => {
  //   const event = {
  //     type: 'click',
  //     target: { id: 'basic-button-0' },
  //     stopPropagation: noop,
  //     preventDefault: noop,
  //   }
  //   jest.spyOn(event, 'stopPropagation')
  //   jest.spyOn(event, 'preventDefault')

  //   const actions = [{ onClick: noop, label: 'Azione 1' }]
  //   const component = renderer.create(
  //     <AllTheProviders>
  //       <ActionMenu index={0} actions={actions}></ActionMenu>
  //     </AllTheProviders>
  //   )
  //   component.root.findByType('button').props.onClick()
  //   const tree = component.toJSON()
  //   expect(tree).toMatchSnapshot()
  // })
})
