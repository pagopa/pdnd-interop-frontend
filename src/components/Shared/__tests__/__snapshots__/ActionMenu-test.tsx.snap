// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot matches button (1) 1`] = `null`;

exports[`Snapshot matches button (2) 1`] = `
<button
  aria-controls="basic-menu-0"
  aria-haspopup="true"
  className="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-1e6y48t-MuiButtonBase-root-MuiButton-root"
  disabled={false}
  id="basic-button-0"
  onBlur={[Function]}
  onClick={[Function]}
  onContextMenu={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={0}
  type="button"
>
  <span>
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeMedium css-ci5apu-MuiSvgIcon-root"
      data-testid="MoreVertIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
      />
    </svg>
  </span>
</button>
`;
